src/experimental-layers/src/mesh-layer/mesh-layer.js

```diff
@@ -0,0 +1,240 @@
+  getAngleDegreesCW: x => x.angle || 0,
+  getColor: x => x.color || [0, 0, 0, 255]
+};
+
+export default class MeshLayer extends Layer {
+
+  getShaders(id) {
+    const {shaderCache} = this.context;
+    return enable64bitSupport(this.props) ?
+      {vs: vs64, fs, modules: ['project64', 'picking', 'lighting'], shaderCache} :
+      {vs, fs, modules: ['picking', 'lighting'], shaderCache}; // 'project' module added by default.
+  }
+
+  initializeState() {
+    const {gl} = this.context;
+    this.setState({model: this.getModel(gl)});
+
+    const {attributeManager} = this.state;
+    attributeManager.addInstanced({
+      instancePositions: {size: 3, accessor: 'getPosition', update: this.calculateInstancePositions},
+      instanceAngles: {size: 1, accessor: 'getAngleDegreesCW', update: this.calculateInstanceAngles},
+      instanceColors: {size: 3, accessor: 'getColor', update: this.calculateInstanceColors}
+    });
+  }
+
+  updateState({props, oldProps, changeFlags}) {
+    super.updateState({props, oldProps, changeFlags});
+
+    const {attributeManager} = this.state;
+
+    if (this.state.dataChanged) {
+      attributeManager.invalidateAll();
+    }
+
+    if (changeFlags.propsChanged) {
+      if (props.fp64 !== oldProps.fp64) {
+        const {gl} = this.context;
+        this.setState({model: this.getModel(gl)});
+      }
+
+      if (props.fp64 && props.coordinateSystem === COORDINATE_SYSTEM.LNGLAT) {
+        attributeManager.invalidateAll();
+        attributeManager.addInstanced({
+          instancePositions64xy: {
+            size: 2,
+            accessor: 'getPosition',
+            update: this.calculateInstancePositions64xyLow
+          }
+        });
+      } else {
+        attributeManager.remove(['instancePositions64xy']);
+      }
+
+      const {sizeScale} = props;
+      this.setUniforms({sizeScale});
+
+      if (props.texture !== oldProps.texture) {
+        if (props.texture) {
+          this.loadTexture(props.texture);
+        } else {
+          // TODO - reset
+        }
+      }
+
+      if (props.lightSettings !== oldProps.lightSettings) {
+        this.setUniforms(props.lightSettings);
+      }
+    }
+  }
+
+  draw({uniforms}) {
+    const {sizeScale} = this.props;
+    this.setUniforms({sizeScale});
+
+    this.state.model.render(Object.assign({}, uniforms, {
+      sizeScale
+    }));
+  }
+
+  getModel(gl) {
+    const isValidMesh =
+      this.props.mesh instanceof Geometry &&
+      this.props.mesh.attributes.positions;
+      // this.props.mesh.attributes.normals &&
+      // this.props.mesh.attributes.texCoords;
+    assert (isValidMesh);
+
+    return new Model(gl, Object.assign({}, this.getShaders(), {
+      id: this.props.id,
+      geometry: this.props.mesh,
+      isInstanced: true
+    }));
+  }
+
+  loadTexture(src) {
+    const {gl} = this.context;
+    const {model} = this.state;
+    getTexture(gl, src).then(texture => model.setUniforms({sampler1: texture}));
```

Comment:
The layer will not rerender when the texture loads because `needsRedraw` is not marked.